oak/server/rust/oak_runtime/src/node/grpc/io.rs

```diff
@@ -0,0 +1,129 @@
+pub struct Sender {
+    runtime: RuntimeProxy,
+    writer: Handle,
+}
+
+impl Sender {
+    pub fn new(runtime: RuntimeProxy, writer: Handle) -> Self {
+        Self { runtime, writer }
+    }
+
+    /// Composes a [`Message`] from `data` and `handles` and sends it to the [`Sender::writer`].
+    pub fn send(&self, data: Vec<u8>, handles: Vec<Handle>) -> Result<(), OakStatus> {
+        self.runtime
+            .channel_write(self.writer, NodeMessage { data, handles })
+    }
+}
+
+/// Wrapper for a [`Handle`] that is responsible for reading messages from an Oak channel.
+/// Uses [`RuntimeProxy`] to read messages from a specific Oak Node.
+pub struct Receiver {
+    runtime: RuntimeProxy,
+    reader: Handle,
+}
+
+impl Receiver {
+    pub fn new(runtime: RuntimeProxy, reader: Handle) -> Self {
+        Self { runtime, reader }
+    }
+
+    /// Waits and reads a [`Message`] from the [`Receiver::reader`].
+    pub fn receive(&self) -> Result<NodeMessage, OakStatus> {
+        let read_status = self.runtime.wait_on_channels(&[self.reader])?;
+
+        if read_status[0] == ChannelReadStatus::ReadReady {
+            self.runtime.channel_read(self.reader).and_then(|message| {
+                message.ok_or_else(|| {
+                    debug!("Channel read error {:?}: Empty message", self.reader);
+                    OakStatus::ErrInternal
+                })
+            })
+        } else {
+            debug!("Channel read error {:?}: {:?}", self.reader, read_status[0]);
+            Err(OakStatus::ErrInternal)
+        }
+    }
+}
+
+/// A gRPC invocation, consisting of exactly two channels: one to read incoming requests from the
+/// client (wrapped in a [`Receiver`]), and one to write outgoing responses to the client (wrapped
+/// in a [`Sender`]).
+pub struct Invocation {
+    receiver: Receiver,
+    sender: Sender,
+}
+
+impl Invocation {
+    /// Reads [`Message`] that contains two channel handles from [`Receiver`] and create
+    /// [`Invocation`] from it.
+    pub fn read(receiver: &Receiver) -> Result<Self, OakStatus> {
+        let message = receiver.receive()?;
+
+        if !message.data.is_empty() {
+            error!(
+                "Incorrect number of bytes received: {} (expected: 0)",
+                message.data.len()
+            );
+            return Err(OakStatus::ErrInternal);
+        }
+        if message.handles.len() != 2 {
+            error!(
+                "Incorrect number of handles received: {} (expected: 2)",
+                message.handles.len()
+            );
+            return Err(OakStatus::ErrInternal);
+        }
+        Ok(Self {
+            receiver: Receiver::new(receiver.runtime.clone(), message.handles[0]),
+            sender: Sender::new(receiver.runtime.clone(), message.handles[1]),
+        })
+    }
+
+    /// Reads an [`GrpcRequest`] from the [`Invocation::receiver`].
+    pub fn receive_request(&self) -> Result<GrpcRequest, OakStatus> {
```

Comment:
I think this method would not be needed if `Sender` and `Receiver` were parametrized.

Reply:
Refactored